base_layer/core/src/transaction_protocol/single_receiver.rs

```diff
@@ -1 +1,149 @@
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+use crate::{
+    range_proof::RangeProof,
+    transaction::{OutputFeatures, TransactionOutput},
+    transaction_protocol::{
+        build_challenge,
+        receiver::RecipientSignedTransactionData as RD,
+        sender::SingleRoundSenderData as SD,
+        TransactionProtocolError as TPE,
+    },
+    types::{CommitmentFactory, PublicKey, SecretKey as SK, Signature, TariCommitment},
+};
+use tari_crypto::keys::PublicKey as PK;
+
+/// SingleReceiverTransactionProtocol represents the actions taken by the single receiver in the one-round Tari
+/// transaction protocol. The procedure is straightforward. Upon receiving the sender's information, the receiver:
+/// * Checks the input for validity
+/// * Constructs his output, range proof and signature
+/// * Constructs the reply
+/// If any step fails, an error is returned.
+pub(super) struct SingleReceiverTransactionProtocol {}
+
+impl SingleReceiverTransactionProtocol {
+    pub fn new(sender_info: &SD, nonce: SK, spending_key: SK, features: OutputFeatures) -> Result<RD, TPE> {
+        SingleReceiverTransactionProtocol::validate_sender_data(sender_info)?;
+        let output = SingleReceiverTransactionProtocol::build_output(sender_info, &spending_key, features)?;
+        let public_nonce = PublicKey::from_secret_key(&nonce);
+        let public_spending_key = PublicKey::from_secret_key(&spending_key);
+        let e = build_challenge(
+            &(&sender_info.public_nonce + &public_nonce),
```

Comment:
why are we using two nonces?

Reply:
One for the sender and one for the receiver